[Spring 체스 - 2단계] 알린(장원영) 미션 제출합니다.

OzRagwort opened the pull request:
안녕하세요 토니! 알린이에요 👋 

이번 미션은 구현도 구현이지만 스프링과 JdbcTemplate이 사용이 익숙하지 않아 문서를 찾아 읽거나 에러를 잡는게 특히 어려웠던 것 같아요..

그래도 열심히 구현해봤습니다. 잘 부탁드립니다! 😃 

이전에 주셨던 피드백에 대한 답은 링크로 남기겠습니다.

[JdbcTemplate을 사용하면서 느낀 차이점](https://github.com/woowacourse/jwp-chess/pull/373#discussion_r857201495)
[생성자 주입을 사용한 이유](https://github.com/woowacourse/jwp-chess/pull/373#discussion_r857202466)

wyRagwort committed:
docs(README): 2단계 요구사항, 1단계 수정 사항 정리

wyRagwort committed:
refactor: 불필요한 의존성 제거

wyRagwort committed:
refactor: 당장 불필요한 메서드 제거

wyRagwort committed:
refactor(BoardDaoImpl): save() 개선

wyRagwort committed:
fix: 말을 잘못 이동한 뒤 보드가 먹통이 되는 에러 수정

wyRagwort committed:
refactor: @Sql을 활용하여 raw한 쿼리문이 테스트에서 노출되지 않도록 수정

wyRagwort committed:
docs: API 기능 정리

wyRagwort committed:
feat: localhost:8080 요청 시 노출되는 페이지와 체스방을 만들 수 있는 버튼 추가

wyRagwort committed:
refactor: 체스방의 정보를 저장할 room 테이블 생성, DAO 메서드 수정

wyRagwort committed:
feat: 체스방 제목과 비밀번호로 새로운 체스판 생성 기능

wyRagwort committed:
refactor: 체스방 고유 번호로 체스방에 접근하고 게임할 수 있도록 수정

wyRagwort committed:
fix: 없는 체스방 URL 요청 시 에러 페이지를 출력하도록 함

wyRagwort committed:
feat: 체스방 목록 조회하기 기능 및 버튼 추가

wyRagwort committed:
feat: 체스방 참여하기 기능

wyRagwort committed:
feat: 체스방 삭제하기 기능 구현

wyRagwort committed:
feat: 체스방 삭제 시 게임이 진행중이면 삭제하지 못하도록 함

wyRagwort committed:
docs: 피드백 답변 정리

wyRagwort committed:
fix: 모든 체스방에서 게임 순서가 제대로 변경되도록 수정

wyRagwort committed:
refactor: 체스방을 만들고 목록을 새로고침하도록 수정

wyRagwort committed:
refactor: 컨벤션 적용

wyRagwort committed:
fix; 체스방을 삭제 시 비밀번호를 입력하도록 수정

toneyparky commented on README.md:
```diff
@@ -1,9 +1,62 @@
-###  기능 요구사항
+### 기능 요구사항
```
탄탄하게 정리해주셨네요! 커밋도 확인해보니 수정사항에 대해서 잘 체크하시며 개발하시는 모습이 인상깊었어요 👍

OzRagwort replied:
정리를 하지않으면 늘 길을 잃어서 간단하게라도 정리하게 되었습니다 

감사합니다 ! 😃

toneyparky commented on src/main/java/chess/web/controller/ChessWebController.java:
```diff
@@ -23,35 +28,59 @@ public ChessWebController(ChessService chessService) {
     @ResponseBody
-    public ResponseEntity<BoardDto> move(@RequestBody MoveDto moveDto) {
-        Board board = chessService.move(moveDto, 1L);
+    public ResponseEntity<BoardDto> move(@RequestBody MoveDto moveDto, @PathVariable Long boardId) {
+        Board board = chessService.move(moveDto, boardId);
         return ResponseEntity.ok().body(BoardDto.from(board));
```
이런 부분에서 에러가 발생할 경우에는 어떻게 처리될까요? 

`ExceptionHandler`와 `ControllerAdvice`라는 두가지 키워드를 학습해보고 적용해보죠!

OzRagwort replied:
`@ControllerAdvice`와 `@ExceptionHandler`를 이용하여 문제가 발생했을 때 에러를 처리할 수 있도록 했습니다.

없는 체스방 URL로 요청 시 체스방 페이지가 아닌 잘못 접속했다는 페이지를 Advice에서 출력하도록 수정했습니다. 

이동할 수 없는 이동을 할 경우 에러를 경고창으로 알려주기위해 Advice에서 예외 메시지를 400에러와 함께 전송하도록 수정했습니다.

OzRagwort replied:
이 부분을 수정하면서 `@Controller`와 `@RestController`의 차이점을 공부하고 적용해봤습니다. 

그래서 기존에 하나였던 ChessWebController를 뷰를 반환하는 ChessWebController와 JSON으로 정보를 주고받는 ChessApiController로 분리해보았습니다.

toneyparky replied:
오오 너무 좋네용! 현재는 백엔드에서 뷰도 내려주는 구조기에 API와 분리해주신거 좋습니다! 컨트롤러의 책임도 명확히 분리되었네요!

toneyparky commented on src/main/java/chess/web/controller/ChessWebController.java:
```diff
@@ -23,35 +28,59 @@ public ChessWebController(ChessService chessService) {
+    public ResponseEntity<Long> createChess(@RequestBody CreateRoomDto createRoomDto) {
+        Long boardId = chessService.createGame();
+        Long newId = chessService.createRoom(boardId, createRoomDto.getTitle(), createRoomDto.getPassword());
+        return ResponseEntity.ok().body(newId);
```
Post 요청이 성공해서 자원이 정상적으로 생성된 경우에는 어떤 status code를 반환할까요?

알린은 status code가 왜 필요하다고 생각하시는지 궁금해요!

OzRagwort replied:
Post 형식으로 요청을 받아서 어떤 데이터가 생성된 경우 statue code(201)이 반환되어야 합니다.

ResponseEntity에서는 `ResponseEntity.created()`를 사용하여 반환할 수 있다.

status code는 요청에 대한 처리 결과가 어떤지 알려주는 코드로 1~5xx 로 반환해줍니다.

각 코드마다 서버나 요청에 대한 상태를 알려주어 사용자가 요청에 대한 결과를 알 수 있게 됩니다.

공부하면서 알게되었는데, 400은 사용자의 실수로 발생하는 에러이고 500은 서버의 문제로 발생하는 에러입니다. 

서버에서는 사용자가 어떤 요청을 보내도 서버의 문제인지 사용자의 문제인지 알려주고 싶지 않아서 500에러는 보통 반환하지 않는다고 합니다. 추가로 어떤 문제인지도 알려주고 싶지 않으면 페이지가 없다는 404 에러코드를 추는 경우도 있다고 하는데 이런 부분이 좀 신기했습니다.

toneyparky replied:
잘 적어주셨네요! API는 프론트와 백 간의 약속이기에 약속을 명확하게 하기 위해서 status code를 사용하죠! 

앞으로 API를 만들 때 마다 status code를 고려한다면 의미가 명확한 약속을 만드실거에요! 

적어주신 500이나 404와 관련된 내용은 저도 알린을 통해서 알게 되었네요! 저는 주로 명확하게 보내주는 편이에요😊

toneyparky commented on src/main/java/chess/web/controller/ChessWebController.java:
```diff
@@ -23,35 +28,59 @@ public ChessWebController(ChessService chessService) {
+    @PostMapping("/chess")
+    public ResponseEntity<Long> createChess(@RequestBody CreateRoomDto createRoomDto) {
+        Long boardId = chessService.createGame();
+        Long newId = chessService.createRoom(boardId, createRoomDto.getTitle(), createRoomDto.getPassword());
```
newId는 새로운 board의 id일까요 room의 id일까요?

OzRagwort replied:
이 문제를 해결하기 위해 확인하던 중 board와 room 테이블이 나뉘어져 있을 이유가 없다는 것을 알게되었습니다. 두 테이블을 합치면서 공통적으로 roomId를 사용하게 되었고 newId도 room 테이블의 pk값이 되었습니다.

toneyparky replied:
좋네요! 항상 기능을 추가하거나 리팩터링을 하면서 기존의 도메인에 수정이 가해진다면 도메인 간의 관계가 망가지지 않았는지 확인하셔야 합니다 😀

toneyparky commented on src/main/java/chess/web/controller/ChessWebController.java:
```diff
@@ -23,35 +28,59 @@ public ChessWebController(ChessService chessService) {
+
+    @DeleteMapping("/chess/{boardId}")
+    public ResponseEntity<Boolean> deleteRoom(@RequestBody String password, @PathVariable Long boardId) {
+        boolean result = chessService.removeRoom(boardId, password);
```
일반적으로 자원을 제거할 때에는 해당 자원의 pk를 이용하여 제거해요. 

현재는 room마다 boardId를 다르게 가져가지만, 데이터의 구조상으로는 제약이 없기에 여러 room이 같은 boardId를 가질 수 있네요.
지금과 같이 boardId로 자원을 제거할 경우에 만일 복수의 room이 같은 boardId를 가진다면 어떨까요? 모든 room을 제거하길 의도하신걸까요?

OzRagwort replied:
board와 room 테이블을 합치면서 board_id가 아닌 roomId로 관리하도록 수정했습니다.

PieceDao의 `deleteByboardId()`는 roomId를 만족하는 모든 기물을 삭제하도록 `deleteAllByRoomId()`로 수정했습니다.

RoomDao는 `delete()`를 사용하여 pk값인 id와 password값을 파라미터로 받아 삭제합니다.

DB에 id와 password가 같은 데이터가 있는지 확인하는 방식(예를들면 `findByIdAndPassword()`)으로 패스워드 일치를 확인하고 `delete(id)`를 하는 식으로 할까 생각했었는데, 글을 삭제하기 위해 무조건 비밀번호가 필요하다고 생각하여 `delete(id, password)`로 DAO 메서드를 만들었습니다.

toneyparky replied:
확인했어요~!

---

delete와 관련해서는 비밀번호 검증과 삭제를 분리하는 데에 장단점이 있을거 같아요. 

지금처럼 id와 password를 동시에 검사할 경우에 id가 없는 경우와 id는 맞지만 password가 틀린 경우를 분리하기 힘들거 같아요. 
만일 지금 상황에서 사용자에게 비밀번호가 틀렸음을 안내하고 싶다면 어떻게 구현할 수 있을까요?

만일 둘을 분리하면 검증과 제거 두번의 쿼리가 날라가기에 효율성이 상대적으로 떨어지겠죠?

요즘 컴퓨팅 자원은 좋기에 저라면 사용성을 위해 둘을 분리할거 같아요 😁

알린의 선택도 좋습니다!

OzRagwort replied:
비밀번호가 틀린경우를 확인하기 위해 쿼리가 두번 날아가더라도 둘을 분리하도록 수정했습니다!

감사합니다 😃

toneyparky commented on src/main/java/chess/web/controller/ChessWebController.java:
```diff
@@ -23,35 +28,59 @@ public ChessWebController(ChessService chessService) {
     }
+
+    @DeleteMapping("/chess/{boardId}")
+    public ResponseEntity<Boolean> deleteRoom(@RequestBody String password, @PathVariable Long boardId) {
```
비밀번호 같은 정보와 식별자를 위한 정보를 각각 body, path variable로 잘 나눠서 보내셨네요! 

웹에 대한 이해도가 돋보입니다😎

OzRagwort replied:
감사합니다 😃

toneyparky commented on src/main/java/chess/web/service/ChessService.java:
```diff
@@ -103,4 +119,45 @@ public ScoreDto getStatus(Long boardId) {
+        }
+        pieceDao.deleteByBoardId(boardId);
+        roomDao.delete(boardId, password);
+        boardDao.deleteById(boardId);
```
만일 이 라인에서 에러가 발생해서 프로그램이 중단된다면 게임의 데이터는 어떤 상태로 유지 될까요?

원자성과 트랜잭션을 키워드로 학습해보고 문제를 해결해볼까요?

OzRagwort replied:
이 문제는 `@Transactional`을 이용하여 해결할 수 있습니다. 

위 코드에서 처럼 내가 정한 모든 작업이 모두 성공하거나 모두 실패해야하는 성질이 원자성입니다. `@Transactional`는 중간에 에러가 발생할 경우 모든 작업을 롤백시켜주기 때문에 모두 성공하거나 모두 실패하게 됩니다. 

또 `@Transactional`을 사용하면 서로 같은 자원을 수정하다가 한쪽이 롤백을 한다거나 할 때 발생하는 문제등을 해결할 수 있는데 이런 성질을 격리성이라고 하는데 `@Transactional`를 사용하면 이런 문제도 해결할 수 있다고 합니다.

`@Transactional`는 여러 비즈니스 로직을 다루는 `Service`나 테스트에서 주로 사용한다고 합니다.

코드에 적용하면서 데이터를 읽기만 하는 경우 `@Transactional(readOnly = true)`으로 트랜잭션 어노테이션을 사용했고 나머지는 기본 설정으로 적용했습니다.

toneyparky replied:
크으 좋네용~

정리 잘 해주셨어요. 현업에서 행위의 트랜잭션을 관리하는건 중요하기에 지금부터 확실하게 연습해두시면 좋습니다!

toneyparky commented on src/main/java/chess/web/service/ChessService.java:
```diff
@@ -103,4 +119,45 @@ public ScoreDto getStatus(Long boardId) {
+                .filter(Piece::isKing)
+                .count();
+
+        return kingCount == 2;
```
2는 어떤 의미인가요?

OzRagwort replied:
게임이 진행되는 동안은 왕이 늘 둘이라 2라고 사용했는데 수정하지 못했네요 ㅠㅠ..

이부분은 기물을 가져와서 Board로 만들고 Board안에서 왕이 죽었는지 판단하는 메서드를 사용하도록 수정했습니다.

toneyparky commented on src/main/java/chess/web/service/ChessService.java:
```diff
@@ -103,4 +119,45 @@ public ScoreDto getStatus(Long boardId) {
+    }
+
+    private boolean matchPassword(Long boardId, String password) {
+        Room room = roomDao.findByBoardId(boardId).get();
```
`findByBoardId` 메서드를 통해 반환된 `Optional`이 빈 값일 경우에는 어떤 일이 발생할까요?

알린은 `Optional`을 사용하는 이유가 뭐라고 생각하세요?

OzRagwort replied:
가장 큰 이유는 null이 올 수 있는 상황을 개발자에게 알려주고 `NullPointerException`을 방지하기 위해서라고 생각합니다.

Optional을 사용하지 않고 null을 처리하는 경우 코드를 작성한 개발자는 알 수 있지만 다른 개발자는 null이 반환될 수 있는 상황인지 알기 어려울 수 있습니다.

하지만 Optional를 사용하면 null이 있을 수 있다는 것을 확실하게 알 수 있고 값이 null인 경우 예외를 발생할지 새로운 값을 만들어 반환할지 등을 정할 수 있습니다.

toneyparky replied:
맞아요 특정 메서드가 그냥 null을 반환하면 해당 메서드를 사용하는 사람은 null이 반환되는지 모르고 null 체크를 안하여 예외가 발생할 수 있기에 이를 명시적으로 드러내준답니다!

다만 Optional을 사용하지만 빈 값인지 체크하지 않고 바로 get을 한다면 예외가 발생할 수 있기에 주의해야 해요.

toneyparky commented on src/main/resources/templates/chess.hbs:
```diff
@@ -0,0 +1,105 @@
+<html lang="ko">
+<head>
+    <meta charset="UTF-8">
+    <title>체스 게임</title>
```
해당 페이지에서 게임이 끝난 후에 목록으로 넘어갈 수 있는 버튼이 있으면 좋겠어요. 

알린이 백엔드를 학습하고 있지만 결국 개발은 사용하는 사람에게 가치를 전달해야하죠. 이를 위해서 사용자를 조금 더 고려하는 법을 배운다면 단순히 기술 이외의 부분에서도 성장할 수 있을거에요!

OzRagwort replied:
더 많은 가치를 만들어 내는 것이 없는지, 불편함은 없는지 등 조금 더 사용자 관점에서 생각해보도록 하겠습니다 

감사합니다 ! 😃

toneyparky reviewed (changes requested):
안녕하세요 알린~! 2단계 미션도 잘 적용해주셨네요!

우선 요구사항을 깔끔하게 구현하셔서 좋았습니다! 웹 뷰도 적절했구요. 

몇가지 수정하면 좋을 사항과 알린이 학습하면 도움이 될 부분에 대해 코멘트 남겼으니 확인 부탁해요😬

화이팅입니다!

wyRagwort committed:
docs: 2단계 수정 사항 정리

wyRagwort committed:
refactor: POST 요청으로 데이터가 생성될 경우 201 status code로 반환하도록 수정

wyRagwort committed:
refactor: board, room 테이블을 합침

wyRagwort committed:
docs: DAO에서 delete를 할 때 pk를 쓰지 않은 메서드에 대한 설명 추가

- pk를 사용한 것은 delete(id)처럼 사용
- pk를 사용하지 않은 것은 조건에 맞는 모든 데이터를 삭제하도록 deleteAllByXXX()처럼 사용

wyRagwort committed:
refactor(Controller): 뷰를 반환하는 컨트롤러와 json을 반환하는 컨트롤러를 분리

wyRagwort committed:
refactor: 발생하는 에러를 `@ExceptionHandler`, `@ControllerAdvice`를 이용하여 처리

wyRagwort committed:
feat: 체스 게임을 하는 페이지에서 목록으로 이동하는 버튼 추가

wyRagwort committed:
refactor: 여러 DAO를 사용하는 메서드에 @Transactional을 사용하도록 수정

wyRagwort committed:
docs: 피드백 답변 정리

toneyparky commented on README.md:
```diff
@@ -1,9 +1,96 @@
+
+---
+
+### 피드백, 수정사항 정리
```
피드백 내용을 정리하는 거 좋네요 👍

OzRagwort replied:
감사합니다! 😃

toneyparky commented on src/main/java/chess/web/controller/ChessApiController.java:
```diff
@@ -0,0 +1,63 @@
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+public class ChessApiController {
```
컨트롤러도 테스트를 해볼까요?

`컨트롤러 테스트를 통해 00000를 검증하겠다`와 같은 테스트 목표를 명확히 세우고 시작해보죠!

OzRagwort replied:
이번에 컨트롤러 테스트를 추가하면서 제가 고민했던 부분입니다. 

1. 값이 제대로 저장, 수정, 호출, 삭제가 되는지 확인하겠다.
2. 올바른 값이 반환되는지 확인하겠다.
3. 적절한 status code가 반환되는지 확인하겠다.
4. 적절한 content type이 사용되는지 확인하겠다.

테스트를 추가하면서 중간중간 제가 생각하지 못했던 부분도 찾을 수 있어서 좋았습니다! 
(게임이 진행중인 경우 예외처리가 아닌 프론트에서 처리하도록 되어있던 점 등)

toneyparky replied:
좋네요! 항상 테스트를 구현할 때에는 검증할 내용을 명확히 해야해요. 그리고 그 내용은 의미가 있어야죠! 알린이 목표하신 네 가지 모두 유의미하네요 👍 

특히 3번만 하고 4까지 안하는 경우가 있는데 꼼꼼하게 처리해주셨네요~!

toneyparky commented on src/main/java/chess/web/controller/ChessApiController.java:
```diff
@@ -0,0 +1,63 @@
+        this.chessService = chessService;
+    }
+
+    @PostMapping(value = "/chess", produces = MediaType.APPLICATION_JSON_VALUE)
```
API는 보통 뷰를 보여주는 엔드 포인트와 구분하여 endpoint에 `/api`를 prefix로 붙이기도 해요!

OzRagwort replied:
구분이 명확하게 될 것 같아서 추가했습니다 😃

toneyparky commented on src/main/java/chess/web/controller/ChessApiController.java:
```diff
@@ -0,0 +1,63 @@
+        return ResponseEntity.created(URI.create("/chess/" + roomId)).body(roomId);
+    }
+
+    @GetMapping(value = "/chess/{boardId}/load", produces = MediaType.APPLICATION_JSON_VALUE)
```
체스 보드라는 자원 하나를 반환하는 API네요! 아래에 남긴 RESTFul API와 관련된 글을 학습해보시고 엔드포인트를 수정해보죠!

OzRagwort replied:
API를 수정하면서 세 가지를 위주로 수정해봤습니다. 

1. 동사형 사용 금지.
2. 적절한 HTTP method 사용.
3. URI에서 의도가 전달이 되는지 확인.

그 결과 아래와 같이 수정되었습니다.

`@Controller`
- GET:`/` : 체스 방 리스트를 확인하고 새로운 체스 방을 만들 수 있는 페이지 출력.
- GET:`/chess/{boardId}` : 체스 게임을 할 수 있는 페이지 출력.

`@RestController`
- POST:`/api/board` : 새로운 체스 게임을 생성한다.
- PUT:`/api/board/{boardId}/initialization` : 체스 게임을 초기화하여 재시작한다.
- PATCH:`/api/board/{boardId}` : 체스 게임에서 시작 위치, 도착 위치를 JSON으로 전송하면 이동한다. 정상적으로 이동 시 수정된 Board를 반환한다.
- GET:`/api/board/{boardId}` : 체스 기물 상태를 가져온다.
- GET:`/api/board/{boardId}/status` : 체스 게임의 블랙, 화이트 팀의 점수를 반환한다.
- DELETE:`/api/board/{boardId}` : 체스 방을 제거한다.

toneyparky replied:
자원 기반으로 잘 정리해주셨네요. 
여기에 추가하자면, board라는 복수의 자원군에서 사용자가 원하는 단일 혹은 복수의 자원을 다룰 수 있게 해주기에 `/api/boards`를 prefix로 수정할거 같아요. RESTFul에서 복수형을 사용하자는 규칙을 강조하기도 해요. 

주머니 안에 공이 여러개 있는 데 그 중에서 하나 혹은 모두 다 줘야할 때를 생각해보면 이해가 쉬우시려나요?!
실제로 주머니 안에 공이 하나가 들어있거나 하나도 없더라도 외부에서 공을 달라고 하는 입장에서는 여러개가 들어 있다고 가정하고 부탁한다고 생각해보죠!

요 부분 수정해볼까요~?!

OzRagwort replied:
수정했습니다!

RESTFul API는 endpoint에서 복수형을 사용하자는 규칙을 강조한다! 기억하겠습니다! 😃

toneyparky commented on src/main/java/chess/web/controller/ChessApiController.java:
```diff
@@ -0,0 +1,63 @@
+        return ResponseEntity.ok().body(BoardDto.from(board));
+    }
+
+    @GetMapping(value = "/chess/{boardId}/status", produces = MediaType.APPLICATION_JSON_VALUE)
```
RESTFul API에서 엔드포인트는 자원을 나타내야 해요. 

그렇기에 알린이 API 를 통해 제공해야 하고픈 자원을 엔드포인트에 드러내도록 수정해보죠! 또한 자원은 여러 개가 있기에 복수로 나타내요. 

저라면 `boards/{boardId}/status` 정도로 표현하여 여러 보드 중에 id를 식별자로하는 보드의 상태라는 자원을 반환한다는 의미를 드러낼거 같아요.

참고 자료 : https://restfulapi.net/

toneyparky commented on src/main/java/chess/web/controller/ChessApiController.java:
```diff
@@ -0,0 +1,63 @@
+        return ResponseEntity.ok().body(status);
+    }
+
+    @PostMapping(value = "/chess/{boardId}/move", produces = MediaType.APPLICATION_JSON_VALUE)
```
자원의 상태를 갱신하는 API는 Post를 사용하는게 맞을까요? Put과 Patch를 학습해보죠!

OzRagwort replied:
Put은 전체를 수정하고 Patch는 일부를 수정할 때 주로 사용한다고 하여 Patch를 사용하도록 수정했습니다.

수정되는 부분은 포지션과 기물의 타입만 바뀌어서 일부만 수정한다고 생각했습니다.

toneyparky replied:
좋네용~ 적절하게 수정해주셨어요

toneyparky commented on src/main/java/chess/web/controller/ChessApiController.java:
```diff
@@ -0,0 +1,63 @@
+        return ResponseEntity.ok().body(BoardDto.from(board));
+    }
+
+    @GetMapping(value = "/chess/{boardId}/restart", produces = MediaType.APPLICATION_JSON_VALUE)
```
해당 엔드포인트는 자원의 상태를 변경시키는군요! Get 으로 처리하는 게 맞을까요?

OzRagwort replied:
이부분은 모든 기물, 순서 등을 모두 초기화하여 수정하기 때문에 Put을 사용하고 `/api/board/{boardId}/initialization`로 수정하였습니다.

toneyparky replied:
Get API을 아무리 찔러도 자원의 상태가 변하지 않는게 기본으로 알고 있어요. 이런 측면에서 초기화한다는 의미를 잘 표현할 수 있는 HTTP  메서드로 잘 수정해주셨네요 👍

toneyparky commented on src/main/java/chess/web/controller/ChessApiController.java:
```diff
@@ -0,0 +1,63 @@
+
+    @DeleteMapping(value = "/chess/{boardId}", produces = MediaType.APPLICATION_JSON_VALUE)
+    public ResponseEntity<Boolean> deleteRoom(@RequestBody String password, @PathVariable Long boardId) {
+        boolean result = chessService.removeRoom(boardId, password);
```
`boardId`가 맞을까요? 해당 변수를 사용하는 `removeRoom `과 같은 곳에서는 `roomId`로 인자를 받네요!

OzRagwort replied:
테이블 구조가 바뀌면서 통일시키지 못했네요 ㅠㅠ

많은 것들이 수정될 때는 특히 더 신경을 써야겠습니다!

toneyparky commented on src/main/java/chess/web/controller/ChessApiControllerAdvice.java:
```diff
@@ -0,0 +1,17 @@
+@RestControllerAdvice
+public class ChessApiControllerAdvice {
+
+    @ExceptionHandler(IllegalArgumentException.class)
+    public ResponseEntity<ErrorResponseDto> illegalArgumentException(IllegalArgumentException exception) {
+        String message = exception.getMessage();
+        return ResponseEntity.badRequest().body(new ErrorResponseDto(Response.SC_BAD_REQUEST, message));
+    }
+}
```
`IllegalArgumentException` 이외의 예외에 대해서는 어떻게 처리할 수 있을까요?

OzRagwort replied:
모든 예외를 처리할 수 있도록 `Exception.class`에 대한 핸들러도 추가했습니다. ChessApiControllerAdvice에서는 모든 예외에 대해 같은 반환이 나온다고 판단하여 `IllegalArgumentException`을 `Exception`으로 수정하였습니다.

toneyparky replied:
Exception.class에 대한 핸들러를 추가해주신 선택 좋아요! 다만 모든 예외를 Exception으로 받고 bad request를 반환하는건 조금 아쉽네요! 

만일 DB나 서버 내부에서 예외가 발생한다해도 사용자는 bad request를 응답 받을테니말이죠!

요 부분 수정해볼까요?!

OzRagwort replied:
맞는지는 모르겠지만 어떤 예외가 발생할지 모르는 상황인데 `Exception`으로 모든 예외를 핸들링하는건 적절하지 않은 것 같다고 느꼈습니다.

그래서 내가 예상할 수 있는 예외에 대해서만 핸들링 하도록 수정했습니다. 

- `SQLException`, `DataAccessException`이 발생하면 500 status code를 반환하도록 수정했습니다.
- `IllegalArgumentException`, `IllegalStateException`, `NoSuchElementException`이 발생하면 400 status code를 반환하도록 수정했습니다.

toneyparky commented on src/main/java/chess/web/dao/RoomDao.java:
```diff
@@ -0,0 +1,18 @@
+
+    Long save(String turn, String title, String password);
+
+    Long updateTurnById(Long id, String turn);
```
해당 메서드를 사용하는 쪽에서 Long 타입 반환값을 사용하나요?

OzRagwort replied:
사용하지 않습니다. 처음에는 수정 후 pk키를 반환하여 수정된 값을 다시 사용할 수 있도록 하려고 했는데 이미 파라미터로 id를 받고있기 때문에 그러지 않아도 될 것 같습니다.

Long 을 void로 수정했습니다!

toneyparky commented on src/test/java/chess/web/dao/RoomDaoTest.java:
```diff
@@ -0,0 +1,92 @@
+    @Sql("/sql/chess-setup.sql")
+    @Test
+    @DisplayName("체스방을 찾는다.")
+    void findByBoardId() {
```
테스트 메서드의 이름과 내용이 맞지 않네요!

OzRagwort replied:
전체적으로 테스트 메서드, DisplayName 등을 확인했습니다!

toneyparky commented on src/test/java/chess/web/dao/RoomDaoTest.java:
```diff
@@ -0,0 +1,92 @@
+        assertThat(optionalRoom.get().getTitle()).isEqualTo(title);
+        assertThat(optionalRoom.get().getPassword()).isEqualTo(password);
+    }
+}
```
깃헙에서 아래와 같이 표시해주네요!

<img width="99" alt="image" src="https://user-images.githubusercontent.com/37579660/166098053-8b9ce5f0-4404-4cab-81a4-cac62fdf4d28.png">


파일의 가장 마지막 줄에 공백이 없다는 의미인데요, 왜 이런 경고문을 보여주는걸까요?

OzRagwort replied:
경고의 내용은 "no newline at end of file"입니다. 이것은 마지막 줄에 공백이 없다는 의미인데 단순한 빈 줄이 아닌 개행 문자가 와야합니다. (\n, \r, \r\n)

마지막 줄에 개행 문자가 있어야하는 이유는 "POSIX"라는 표준에서 그렇게 정했기 때문입니다. 

많은 시스템이나 프로그램이 이 표준에 맞추어 구현되어있기 때문에 마지막 줄에 개행 문자가 없으면 의도하지 않은 동작이 발생할 수 있다고 합니다.

예를 들어 `gcc`의 경우 POSIX에 근거하여 마지막 줄에 개행 문자가 없으면 정상적으로 동작하지 않기도 한다고 합니다.

또 `cat`의 경우 실제로 프로그램의 끝까지 출력하고 다시 명령을 받기 위해 쉘 입력창이 출력될 때 마지막 출력 이후 바로 입력을 받기 때문에 개행 문자가 있는 것이 읽기에 더 좋다고 합니다. 😄

toneyparky replied:
정리 잘 해주셨어요 알린! 말씀해주신대로 의도하지 않은 동작을 방지하기 위해서 마지막 공백 라인을 넣어주는 게 좋아요. 사람이 확인하는 것도 좋지만 코딩 스타일 규칙으로 설정해두면 더 좋겠죠?

OzRagwort replied:
인텔리제이에서 Preferences - Editor - General - On Save - Ensure every saved file ends with a line break를 체크하여 설정해줬습니다.

제 인텔리제이는 한글이라 환경설정 - 에디터 - 일반 - 저장 시 - 모든 저장된 파일이 줄 바꿈으로 끝나도록 함. 을 설정해주었습니다!

toneyparky commented on src/main/java/chess/web/service/ChessService.java:
```diff
@@ -1,106 +1,169 @@
+            return false;
+        }
+        Room room = optionalRoom.get();
+        if (isRunningChess(roomId, room) || !matchPassword(room, password)) {
+            return false;
+        }
+        return true;
```
```suggestion
        return !isRunningChess(roomId, room) && matchPassword(room, password);
```

아래와 같이 바꿀 수 있겠네요!

해당 라인을 메서드로 추출하면 더 가독성이 좋아지겠죠?

OzRagwort replied:
네 수정했습니다!

toneyparky commented on src/main/java/chess/web/service/ChessService.java:
```diff
@@ -1,106 +1,169 @@
+    }
+
+    private boolean canRemoveRoom(Long roomId, String password) {
+        Optional<Room> optionalRoom = roomDao.findById(roomId);
```
Type으로 Optional인지 알려주기에 변수명으로는 표현하지 않아도 될거 같아요.

OzRagwort replied:
생각해보니 그렇네요 String 변수라고 변수명에 string이 들어가지 않는 것 처럼 생각하니 이해가 쉬웠습니다.

감사합니다!

toneyparky commented on src/main/java/chess/web/service/ChessService.java:
```diff
@@ -1,106 +1,169 @@
+
+    private boolean canRemoveRoom(Long roomId, String password) {
+        Optional<Room> optionalRoom = roomDao.findById(roomId);
+        if (optionalRoom.isEmpty()) {
+            return false;
+        }
```
삭제하려는 방이 없을 경우에 false를 반환하는군요! 저라면 예외를 던지고 프론트에서 예외가 왔을 경우에 사용자에게 적절한 정보를 보여줄거 같아요! 

이렇게 수정해볼까요?

OzRagwort replied:
`NoSuchElementException`을 던지도록 수정했습니다!

toneyparky commented on src/main/java/chess/web/service/ChessService.java:
```diff
@@ -1,106 +1,169 @@
+        Pieces pieces = Pieces.from(pieceDao.findAllByRoomId(roomId));
+        Board board = Board.create(pieces, turn);
+
+        boolean deadKing = board.isDeadKing();
+        return !deadKing;
```
저라면 아래와 같이 수정할거 같아요.

```suggestion
        return board.isKingAlive();
```

제가 작성한 코드와 알린이 작성한 코드는 어떤 차이가 있을까요? 저는 두가지 정도가 생각나네요!

OzRagwort replied:
제가 고민한 차이점입니다. 

1. isXXX()메서드를 사용하면서 영향을 주는 메서드들의 결을 맞추기 위해. 추가로 부정(!) 사용을 최소화하기 위해.
2. 저는 두 줄로 나누어 코드를 작성했는데 한줄로 줄일 수 있을 것 같다.

- is, has, can 등 boolean형을 반환하는 메서드는 가능하면 연관된 여러 메서드들의 결을 통일시키는 것이 좋다고 들었습니다.
- 예를 들면 연료가 있어야 자동차가 이동할 수 있다면 자동차가 이동할 수 있는지 확인하는 canMoveCar()라는 메서드에서 hasFuel()이라는 메서드를 쓰는것이 좋다는 의미입니다.
- 이런 통일성을 맞추기위 위한 기준은 보통 긍정형으로 메서드를 만드는 것이 좋다고 들었습니다.
- 예를들면 isNotKing()보다 isKing()이 좋다는 의미입니다.
- 이런 관점에서 `isRunningChess()`메서드에서는 `!isDeadKing()`과 `isKingAlive()`중 `isKingAlive()`를 사용하는 것이 더 좋아보입니다. 왕이 살아있어야 게임이 진행중이기 때문입니다.
- `isRunningChess()`와 `isDeadKing()`메서드의 특징이 다르기 때문에 반환을 하면서 !를 붙여줘야하고 코드의 이해도가 낮아질 것 같습니다.
- 심지어 프로덕션 코드에서 `if (!isRunningChess())`와 같이 사용되기 때문에 부정을 두번 하고 있어 더 이해도가 낮아질 것 같습니다.
- 이 문제를 해결하기 위해 `isKingAlive()`를 만들어 사용하거나 `isRunningChess()`를 `isFinishedChess()`로 수정하고 `isDeadKing()`를 반환하는 것도 방법이 될 것 같습니다.
- 저는 새로운 Board에 새로운 메서드를 만들지 않아도 되고 상태에 대한 확인이기 때문에 `isRunningChess()`를 `isFinishedChess()`로 수정하는 방식으로 수정해봤습니다.

toneyparky replied:
좋네요~! 제가 제안한 리뷰에 추가로 `isFinishedChess`로 수정하셨다는 내용을 보고 알린의 고민이 느껴졌습니다 😁
코드를 읽는 사람을 배려한 좋은 방향성이라 생각해요 👍 

적어주신 두가지 차이점도 제가 생각했던 내용과 동일해요. 2번을 조금 더 발전시키면 필요 없는 변수는 할당하지 말자 정도가 되겠네요.

toneyparky reviewed (changes requested):
안녕하세요 알린, 토니에용~

리뷰를 빠르게 반영해주셔서 좋네요! 게다가 꼼꼼하게 학습하시고 댓글도 남겨주셔서 답변하는 재미가 있었습니다!

수정하면 좋을 내용과 질문을 다시 남겼으니 확인 부탁해요~~

남은 하루도 즐겁게 보내세용😎

wyRagwort committed:
docs: 2단계 추가 수정 사항 정리

wyRagwort committed:
refactor: 사용하지 않는 용어 boardId를 roomId로 수정

wyRagwort committed:
refactor(RoomDao): updateTurnById()의 반환형을 Long에서 void로 수정

wyRagwort committed:
refactor: 변수명 수정, 코드 가독성 개선

wyRagwort committed:
refactor: 파일의 가장 마지막 줄에 개행 문자 추가 및 이유 정리

wyRagwort committed:
refactor(ControllerAdvice): 특정 예외를 제외한 다른 에러도 처리할 수 있도록 수정

wyRagwort committed:
refactor: 체스방 삭제 시 체스방이 없는 경우 에러를 출력하도록 수정

wyRagwort committed:
refactor: 체스방 삭제 시 id는 맞지만 비밀번호가 틀린 경우 에러를 출력하도록 수정

wyRagwort committed:
refactor: boolean 반환 메서드 가독성이 좋아지도록 수정

wyRagwort committed:
refactor: API endpoint에 prefix로 "/api"를 추가

wyRagwort committed:
refactor: RESTFul API에 맞게 endpoint를 수정

wyRagwort committed:
refactor: Test를 할 때 H2를 사용하도록 테스트용 설정 파일 생성

wyRagwort committed:
test: Controller 테스트 추가

wyRagwort committed:
refactor: 잘못된 테스트, 클래스 명, 컨벤션 수정

wyRagwort committed:
refactor: 잘못된 변수명 수정

OzRagwort commented:
안녕하세요 토니 알린이에요 리뷰 요청 드립니다!

꼼꼼하고 깊은 고민을 하게 하는 피드백이 많아서 좋았습니다! 

그리고 많은 칭찬을 해주셔서 너무 기분이 좋아요 😄 

감사합니다 토니! 👍

wyRagwort committed:
fix: 체스 목록을 요청하면 비밀번호가 전송되지 않도록 수정

toneyparky commented on docker/db/mysql/init/init.sql:
```diff
@@ -1,15 +1,15 @@
-    id   INT(10) not null AUTO_INCREMENT ,
+    id  INT(10) not null AUTO_INCREMENT,
     turn VARCHAR (5) not null,
+    title VARCHAR(30) not null,
+    password VARCHAR(30) not null,
```
30자 이상의 이름과 비밀번호를 가지는 방을 생성할 경우 사용자는 어떤 경험을 하게 될까요?!

OzRagwort replied:
결과적으로 값은 저장되지 않고 `DataIntegrityViolationException`이 발생했습니다. 이 예외는 `DataAccessException`를 상속받는 예외입니다.

이 상황 말고도 데이터에 잘못 접근하거나 처리할 때 `DataAccessException`예외가 발생했는데 보통은 사용자가 잘못된 값을 입력하거나 잘못된 데이터를 받아오면서 발생했기 때문에 400에러를 발생시켜도 된다고 생각했습니다.

제가 생각한 상태 코드는 400이지만 혹시 수정이 될 수 있다고 생각하여 Advice클래스에서 메서드로 분리했습니다.

toneyparky replied:
저도 400 에러를  보낼거 같아요! 적절하게 처리해주셨네요 😊

toneyparky commented on src/main/java/chess/web/controller/ChessApiController.java:
```diff
@@ -0,0 +1,67 @@
+        return ResponseEntity.ok().body(BoardResponseDto.from(board));
+    }
+
+    @GetMapping(value = "/board/{boardId}/status", produces = MediaType.APPLICATION_JSON_VALUE)
```
해당 API도 GET 메서드인데 프로그램 내부의 상태를 바꾸는군요?

OzRagwort replied:
`loadGame()`에서는 board만 반환하고 어떠한 상태를 변화하는 모든 로직을 지웠습니다.

더 확실하게 관리하기 위해 `@Transactional`의 `readOnly = true`옵션을 추가해주었습니다.

제가 확인하지 못한 부분까지 확인해주셔서 감사합니다 👍

toneyparky replied:
API가 보다 명확해졌어요! 고생하셨습니다.

toneyparky commented on src/main/java/chess/web/controller/ChessApiController.java:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    @DeleteMapping(value = "/board/{boardId}", produces = MediaType.APPLICATION_JSON_VALUE)
+    public ResponseEntity<Boolean> deleteRoom(@RequestBody String password, @PathVariable Long boardId) {
```
room의 흔적이 있네요!

Room => BoardEntity 라는 선택을 하신 이유는 프로젝트의 요구사항을 표현할 보다 적절한 명칭이라 생각하셔서일까요?

OzRagwort replied:
예전에 VO에 대해 공부하면서 VO, DTO, Entity를 비교하는 글을 본적이 있습니다.

서로 조금씩 특징이 달랐는데 제가 느낀 Entity는 데이터베이스의 값을 그대로 매핑시켜 포장해둔 클래스였습니다. 그래서 어떤 로직이 있지는 않고 원하는 데이터를 사용하기 위해서는 도메인이나 DTO로 바꾸고 사용했습니다. 

삭제 부분에서도 삭제를 하려면 비밀번호를 확인해야 합니다. 그래서 비밀번호까지 가진 BoardEntity로 값을 받아와서 검사하고, 서버로 보드의 정보를 보낼때는 비밀번호가 없는 BoardResponseDto로 만들어 전송하도록 수정했습니다.

toneyparky replied:
크.. 좋네요. BoardResponseDto 만들어주신 부분이 인상깊어요!

toneyparky commented on src/main/java/chess/web/controller/ChessWebControllerAdvice.java:
```diff
@@ -0,0 +1,21 @@
+public class ChessWebControllerAdvice {
+
+    @ExceptionHandler(NoSuchElementException.class)
+    public String hasNotChessRoom(NoSuchElementException exception) {
```
이 친구도 Room의 개념이 남아있네요~!

OzRagwort replied:
command + shift + F 로 전체 프로젝트에 값을 찾아서 수정했었는데 제가 대소문자 구분을 하면서 "Room"이라고 적혀있는 부분을 수정하지 못한 것 같아요. ㅠㅠ

조금 더 신경써야겠습니다..! 😢

toneyparky commented on src/test/java/chess/web/controller/ChessApiControllerAdviceTest.java:
```diff
@@ -0,0 +1,145 @@
+        List<Piece> savedPieces = chessService.loadGame(id).getPieces().getPieces();
+        Optional<Piece> pieceByFromPosition = findByPosition(savedPieces, Position.from(fromPosition));
+        Optional<Piece> pieceByToPosition = findByPosition(savedPieces, Position.from(toPosition));
+        assertThat(pieceByFromPosition).isPresent();
+        assertThat(pieceByToPosition).isPresent();
+        assertThat(pieceByFromPosition.get()).isInstanceOf(Pawn.class);
+        assertThat(pieceByToPosition.get()).isInstanceOf(Empty.class);
```
assertAll 이라는 검증 함수를 사용하면 어떤 장점을 얻을 수 있을까요?

OzRagwort replied:
`assertAll()`는 모든 테스트가 실행될 때 까지 예외를 발생시키지 않아 모든 테스트를 우선은 검증한다는 의미입니다.

`assertThat(A)`, `assertThat(B)`, `assertThat(C)` 순서대로 검증을 하고 있다가 만약 `assertThat(A)`에서 실패한다면 B와 C는 검증조차 하지않게되는데 이런 경우 B, C 테스트는 A 테스트에 의존하게 됩니다.

불필요하게 검증들이 서로 의존하지 않도록 하기위해 `assertAll()`을 사용합니다.

toneyparky replied:
네 적어주셨듯 단언이 실패하더라도 다른 단언에 영향을 끼치지 않도록 하기 위해서 해당 함수를 사용합니다!

toneyparky commented on src/main/java/chess/web/service/ChessService.java:
```diff
@@ -103,4 +121,58 @@ public ScoreDto getStatus(Long boardId) {
+        return true;
+    }
+
+    private boolean canRemoveRoom(Long boardId, String password) {
```
Room 개념이 남아있습니다 😁

OzRagwort replied:
수정했습니다 ! 😭

toneyparky reviewed (changes requested):
안녕하세요 알린! 좋은 아침이에요. 

리뷰 반영 잘 해주셔서 감사해요. 고민을 많이 하신게 느껴졌어요. 

머지할까 했지만 아직 시간이 남았기에 몇가지 추가로 리뷰 드렸습니다! 확인 및 반영 부탁드려요~!

wyRagwort committed:
docs: 2단계 추가 수정사항 정리

wyRagwort committed:
refactor: REST API의 endpoint의 자원 복수형으로 수정

wyRagwort committed:
refactor: 제거된 테이블인 Room의 흔적을 제거

wyRagwort committed:
refactor: 30자 이상의 방 이름, 비밀번호를 입력할 때도 Advice를 할 수 있도록 수정

wyRagwort committed:
refactor: Advice에서 Exception을 핸들링하지 않도록 수정

wyRagwort committed:
refactor: GET으로 요청할 때 상태를 변화시키지 않도록 수정

wyRagwort committed:
refactor(test): 여러 검증을 하는 경우 assertAll()를 사용하도록 수정

OzRagwort commented:
안녕하세요 토니 ! 

늘 좋은 리뷰와 응원해주셔서 감사합니다. 

제가 신경쓰지 못한 부분까지 확인해주시는 세심함에 감동했습니다 !

토니 늘 행복하세요! 😄

toneyparky reviewed (approved):
알린~ 리뷰 반영해주신 내용 확인했어요! 잘 작성해주셨네요 👍

이번 미션을 통해 웹과 스프링에 대해서 충분히 학습하셨길 바라며 머지하도록 할게요 😁

고생 많으셨어요!
